fdbclient/BackupAgentBase.actor.cpp

```diff
@@ -739,8 +782,36 @@ ACTOR Future<int> kvMutationLogToTransactions(Database cx,
 				                          cx,
 				                          tenantMap,
 				                          provisionalProxy));
+				if (tenantMapChanging && req.transaction.mutations.size()) {
+					// If the tenantMap is changing send the previous CommitTransactionRequest to the CommitProxy
+					TraceEvent("MutationLogRestoreTenantMapChanging").detail("BeginVersion", newBeginVersion);
+					CODE_PROBE(true, "mutation log tenant map changing");
+					wait(sendCommitTransactionRequest(req,
+					                                  uid,
+					                                  newBeginVersion,
+					                                  rangeBegin,
+					                                  committedVersion,
+					                                  &totalBytes,
+					                                  &mutationSize,
+					                                  addActor,
+					                                  commitLock,
+					                                  commit));
+					req = CommitTransactionRequest();
+					mutationSize = 0;
+				}
+
+				state int i;
```

Comment:
This logic does not seem quite aligned with what we fundamentally need.

I think what we actually need is:  "Any time a transaction batch with a metadata mutation is added to the current CTR being accumulated, after the addition commit the current CTR and then reset the current CTR."  It seems easiest to me to reason about the correctness of this implementation.

What is being done here instead is committing a CTR _before_ a new batch addition which modifies tenant metadata is added to the current CTR, which is actually 1 step early as there is no dependency problem to have `[user mutations only], [user mutations only], ... [user mutations and a tenant change] <COMMIT HERE>`.

I can't tell if there is an edge case bug where you have these batches from logValue decoding:
```
1. [mutations]
2. [mutations + create tenant A
3. [mutations using tenant A but no tenant metadata changes]
```
I think it works, because a CTR flush of (1) would be forced before (2) and then the loop would break and then flush (2) alone without ever reading (3) and then exit the function.  The next function start would then try to flush a CTR before (3) but it would skip the flush because the current CTR would be empty at that point, then it would continue reading.

Basically it seems like the `break` on line 815 is what is _actually_ providing the "flush CTR after a tenant change" that we actually require here, in which case this loop can be simplified.

Also, it's not clear to me how this function is able to break the read loop based on hitting the max BATCH_SIZE (the existing logic which modified to also stop on tenant metadata change) because the caller of this function calls it just once for each range in a range result set:
```
			for (idx = 0; idx < ranges.size(); ++idx) {
				int bytes =
				    wait(kvMutationLogToTransactions(cx,
				                                     results[idx],
				                                     locks[idx],
```
No `idx` is repeated, so if this actor returns before processing everything in the input range then it seems like mutations would be just skipped.  Since the batch size limit is only 1MB I would expect this to cause actual test failures if it wasn't correct, but I do not understand why it is correct.

Reply:
From a discussion offline, the requirement I was missing here is that the CommitProxy downstream will not allow the type of transaction I describe in (2) above, so it is required that we force a commit both _before_ and _after_ any original commit from a `logValue` which contains a tenant metadata change.

Reply:
I am still wondering how the `idx` loop actually processes all of the range results completely given that the actor can exit early and the outer loop will move on to the next `idx`.  Theoretically if Backup/Restore tests do not make commits > 1MB then this could be a bug we just never caught.

Reply:
Disregard, I missed the outer `loop` here on line 754!